src/circuit/dbsp_handle.rs

```diff
@@ -0,0 +1,309 @@
+        // to us via this channel.  The function returns after receiving a
+        // notification from each worker.
+        let (init_senders, init_receivers): (Vec<_>, Vec<_>) =
+            (0..nworkers).map(|_| bounded(0)).unzip();
+
+        // Channels used to send commands to workers.
+        let (command_senders, command_receivers): (Vec<_>, Vec<_>) =
+            (0..nworkers).map(|_| bounded(1)).unzip();
+
+        // Channels used to signal command completion to the client.
+        let (status_senders, status_receivers): (Vec<_>, Vec<_>) =
+            (0..nworkers).map(|_| bounded(1)).unzip();
+
+        let runtime = Self::run(nworkers, move || {
+            let worker_index = Runtime::worker_index();
+
+            // Drop all but one channels.  This makes sure that if one of the worker panics
+            // or exits, its channel will become disonnected.
+            let init_sender = init_senders.into_iter().nth(worker_index).unwrap();
+            let status_sender = status_senders.into_iter().nth(worker_index).unwrap();
+            let command_receiver = command_receivers.into_iter().nth(worker_index).unwrap();
+
+            let circuit = match Circuit::build(constructor) {
+                Ok((circuit, res)) => {
+                    if init_sender.send(Ok(res)).is_err() {
+                        return;
+                    }
+                    circuit
+                }
+                Err(e) => {
+                    let _ = init_sender.send(Err(e));
+                    return;
+                }
+            };
+
+            // TODO: uncomment this when we have support for background compaction.
+            // let mut moregc = true;
+
+            while !Runtime::kill_in_progress() {
+                // Wait for command.
+                match command_receiver.try_recv() {
+                    Ok(()) => {
+                        //moregc = true;
+                        let status = circuit.step();
+                        // Send response.
+                        if status_sender.send(status).is_err() {
+                            return;
+                        };
+                    }
+                    // Nothing to do: do some housekeeping and relinquish the CPU if there's none
+                    // left.
+                    Err(TryRecvError::Empty) => {
+                        // GC
+                        /*if moregc {
+                            moregc = circuit.gc();
+                        } else {*/
+                        Runtime::parker().with(|parker| parker.park());
+                        //}
+                    }
+                    Err(_) => {
+                        break;
+                    }
+                }
+            }
+        });
+
+        // Receive initialization status from all workers.
+
+        let mut init_status = Vec::with_capacity(nworkers);
+
+        for (worker, receiver) in init_receivers.iter().enumerate() {
+            match receiver.recv() {
```

Comment:
I guess it's not an issue, but a consequence here of having a separate send channel per worker here is that we're waiting/blocking on each worker serially until all have started (as opposed to handling the data as it is produced).